### PR TITLE
Make AnimationMixer consider Discrete for RESET track

### LINKED ISSUE #89163: Godot 4.3.dev4 bug - animation on AnimatedSprite2D not working if frame is present in RESET animation
### Tested versions

- Reproducible in Godot 4.3.dev4.official
- Not reproducible in Godot 4.3.dev3.official

### System information

Godot v4.3.dev4 - Windows 10.0.19045 - GLES3 (Compatibility) - NVIDIA GeForce GTX 850M (NVIDIA; 31.0.15.3168) - Intel(R) Core(TM) i7-4710HQ CPU @ 2.50GHz (8 Threads)

### Issue description

AnimatedSprite2D animations not playing based on AnimationPlayer setup.

Consider this setup:
1. An AnimatedSprite2D called X with a couple of animations that loop.
2. An AnimationPlayer with animations that only change the "animation" property of X.
3. An AnimationTree that changes the animations based on a variable that changes based on user input.

I found that, if the AnimationPlayer's "RESET" animation contains the "frame" property of X, then the animations don't play. If you remove the "frame" property then they run fine. Please see the MRP attached.

This used to work in Godot 4.2 and 4.3.dev3, so this is a regression in behavior in 4.3.dev4.

### Steps to reproduce

Described above.

### Minimal reproduction project (MRP)

[BugAnimationPlayer.zip](https://github.com/godotengine/godot/files/14488594/BugAnimationPlayer.zip)

I included 2 scenes, one where the animations don't play, and one where they play fine. The only difference is one has the "frame" property in the RESET animation, and the other does not.

### PR DESCRIPTION
**Pull request by TokageItLab**
- Fixes #89163

Deterministic sets the initial value if an animation are lacking some tracks from another animation, but this should not be done for Discrete tracks, so add some flagging when caching and initializing. It should behave as follows:

| Deterministic  | RESET Track | CallbackModeDiscrete | Is init value appeared in animation with lacking some tracks |
| -- | -- | -- | -- |
| true | Continuous | Dominant | Yes, init value is continuous clearly |
| true | Continuous | Recessive | Yes, init value is continuous clearly |
| true | Continuous | Force Continuous | Yes, init value is continuous clearly |
| true | Discrete| Dominant | No, init value is discrete clearly |
| true | Discrete| Recessive | No, init value is discrete clearly |
| true | Discrete | Force Continuous | No, init value is discrete clearly |
| true | None (but there is track in another animation)  | Dominant | No, init value may be discrete |
| true | None (but there is track in another animation)  | Recessive | No, init value may be discrete |
| true | None (but there is track in another animation)  | Force Continuous | Yes, init value is discrete clearly by Force Continuous |
| false | Continuous | Dominant | No, Un-Deterministic doesn't process init value |
| false | Continuous | Recessive | No, Un-Deterministic doesn't process init value |
| false | Continuous | Force Continuous | No, Un-Deterministic doesn't process init value |
| false | Discrete | Dominant | No, Un-Deterministic doesn't process init value |
| false | Discrete | Recessive | No, Un-Deterministic doesn't process init value |
| false | Discrete | Force Continuous |  No, Un-Deterministic doesn't process init value |
| false | None (but there is track in another animation)  | Dominant | No, Un-Deterministic doesn't process init value |
| false | None (but there is track in another animation)  | Recessive | No, Un-Deterministic doesn't process init value |
| false | None (but there is track in another animation)  | Force Continuous | No, Un-Deterministic doesn't process init value |

### CODE REVIEW

File: scene/animation/animation_mixer.cpp [hunk 1]
```diff
@@ -1975,6 +1982,7 @@ void AnimationMixer::_build_backup_track_cache() {
 				if (t_obj) {
 					t->value = Animation::cast_to_blendwise(t_obj->get_indexed(t->subpath));
 				}
+				t->use_continuous = true;
```
Comment (lyuma) [inline scene/animation/animation_mixer.cpp:1985]: Why is use_continuous necessarily true in `_build_backup_track_cache`? There seems to be no check here: What if the track was discrete?

File: scene/animation/animation_mixer.cpp [hunk 2]
```diff
@@ -696,14 +696,19 @@ bool AnimationMixer::_update_caches() {
 						track_value->init_value = anim->track_get_key_value(i, 0);
 						track_value->init_value.zero();
 
+						track_value->init_use_continuous = callback_mode_discrete == ANIMATION_CALLBACK_MODE_DISCRETE_FORCE_CONTINUOUS;
```
Comment (lyuma) [inline scene/animation/animation_mixer.cpp:699]: Why does the value of callback_mode_discrete at initialization time affect its blending later?

Line 1428 will use this to calculate `force_continuous` and force `use_continuous = true` at that time.

File: scene/animation/animation_mixer.cpp [hunk 3]
```diff
@@ -1733,7 +1740,7 @@ void AnimationMixer::_blend_apply() {
 			case Animation::TYPE_VALUE: {
 				TrackCacheValue *t = static_cast<TrackCacheValue *>(track);
 
-				if (!t->is_variant_interpolatable || (callback_mode_discrete == ANIMATION_CALLBACK_MODE_DISCRETE_DOMINANT && t->use_discrete)) {
+				if (!t->is_variant_interpolatable || !t->use_continuous || (callback_mode_discrete == ANIMATION_CALLBACK_MODE_DISCRETE_DOMINANT && t->use_discrete)) {
```
Comment (lyuma) [inline scene/animation/animation_mixer.cpp:1743]: Is the reason t hat if use_continuous was not initialized to `callback_mode_discrete == ANIMATION_CALLBACK_MODE_DISCRETE_FORCE_CONTINUOUS` here, then it will need to be

`(!!t->use_continuous && callback_mode_discrete != ANIMATION_CALLBACK_MODE_DISCRETE_FORCE_CONTINUOUS)`

but won't line 1430 have always run before hitting this?

**Review by lyuma — APPROVED** [review]
I think I see the idea, that this code wishes to add an additional case to disallow continuous blending if there is no continuous RESET track but not in a force continuous blend.

When there is no RESET track for a particular value, I understand it seems to default to true, but I don't understand why it is forced to true when FORCE_CONTINUOUS is set

Anyway I'll mark it approved because I trust your judgement and I'm not sure anyone is familiar enough with the track cache to give a better review.

Basically, this is introducing additional state to the track cache, and there are more cases that set it to true than set it false, so I want to make sure it is correct.

File: scene/animation/animation_mixer.cpp (hunk 1 above)
Comment (TokageItLab) [reply scene/animation/animation_mixer.cpp:1985]: In the case of discrete, values may not be processed depending on the key position, and the scene values may be rewritten when saving.

File: scene/animation/animation_mixer.cpp (hunk 2 above)
Comment (TokageItLab) [reply scene/animation/animation_mixer.cpp:699]: As shown in the table in the description, this is because the process needs to be changed depending on the presence or absence of the reset track.

File: scene/animation/animation_mixer.cpp (hunk 3 above)
Comment (TokageItLab) [reply scene/animation/animation_mixer.cpp:1743]: At the time of blending about line 1430, it is not yet possible to determine if discrete is present there or not. Once iteration is complete, it is necessary to determine if the result should really be applied.

**Review by lyuma — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!
